rustls/src/client/ech.rs

```diff
@@ -0,0 +1,509 @@
+#![allow(dead_code)] // TODO(@cpu): remove in subsequent commit.
+use alloc::boxed::Box;
+use alloc::vec;
+use alloc::vec::Vec;
+
+use pki_types::{DnsName, ServerName};
+use subtle::ConstantTimeEq;
+
+use crate::client::EchConfig;
+use crate::crypto::hash::Hash;
+use crate::crypto::hpke::{EncapsulatedSecret, HpkeProvider, HpkePublicKey, HpkeSender, HpkeSuite};
+use crate::crypto::SecureRandom;
+use crate::hash_hs::{HandshakeHash, HandshakeHashBuffer};
+#[cfg(feature = "logging")]
+use crate::log::trace;
+use crate::msgs::base::{Payload, PayloadU16};
+use crate::msgs::codec::Codec;
+use crate::msgs::enums::{EchVersion, ExtensionType, HpkeAead, HpkeKdf, HpkeKem};
+use crate::msgs::handshake::{
+    ClientExtension, ClientHelloPayload, EchConfig as EchConfigMsg, EchConfigContents,
+    EncryptedClientHello, EncryptedClientHelloOuter, HandshakeMessagePayload, HandshakePayload,
+    HelloRetryRequest, HpkeKeyConfig, HpkeSymmetricCipherSuite, Purpose, Random,
+    ServerHelloPayload, SessionId,
+};
+use crate::msgs::message::{Message, MessagePayload};
+use crate::tls13::key_schedule::{server_ech_hrr_confirmation_secret, KeyScheduleHandshakeStart};
+use crate::{
+    AlertDescription, CommonState, Error, HandshakeType, PeerIncompatible, PeerMisbehaved,
+    ProtocolVersion, Tls13CipherSuite,
+};
+
+/// An enum representing ECH offer status.
+pub(crate) enum Status {
+    /// ECH was not offered - it is a normal TLS handshake.
+    NotOffered,
+    /// ECH was offered and the server accepted.
+    Accepted,
+    /// ECH was offered and the server rejected.
+    Rejected,
+}
+
+/// Randomly generated X25519 public key to use for GREASE ECH HPKE.
+const DUMMY_25519_PUBKEY: &[u8] = &[
+    0x67, 0x35, 0xCA, 0x50, 0x21, 0xFC, 0x4F, 0xE6, 0x29, 0x3B, 0x31, 0x2C, 0xB5, 0xE0, 0x97, 0xD8,
+    0xD0, 0x58, 0x97, 0xCF, 0x5C, 0x15, 0x12, 0x79, 0x4B, 0xEF, 0x1D, 0x98, 0x52, 0x74, 0xDC, 0x5E,
+];
```

Comment:
Style nit: these should go to the bottom.

Reply:
Fixed.